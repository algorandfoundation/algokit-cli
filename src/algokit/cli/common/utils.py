# Cli/Click related helper functions and classes


import typing as t

import click

from algokit.cli.common.constants import ExplorerEntityType


class MutuallyExclusiveOption(click.Option):
    """
    A Click option that defines mutually exclusive command line options.

    Args:
        *args: Positional arguments passed to the parent class constructor.
        **kwargs: Keyword arguments passed to the parent class constructor.
            not_required_if (list): A list of options that the current option is mutually exclusive with.

    Attributes:
        not_required_if (list): A list of options that the current option is mutually exclusive with.

    Raises:
        AssertionError: If the `not_required_if` parameter is not provided.

    Example:
        ```python
        @click.command()
        @click.option('--option1', help='Option 1')
        @click.option('--option2', help='Option 2')
        @click.option('--option3', help='Option 3', cls=MutuallyExclusiveOption, not_required_if=['option1', 'option2'])
        def my_command(option1, option2, option3):
            # Command logic here
            pass
        ```

        In the example above, the `MutuallyExclusiveOption` class is used to define the `option3` command line option.
        This option is mutually exclusive with `option1` and `option2`,
        meaning that if either `option1` or `option2` is provided, `option3` cannot be used.
        If `option3` is provided along with `option1` or `option2`, a `click.UsageError` is raised.
    """

    def __init__(self, *args: t.Any, **kwargs: t.Any) -> None:
        self.not_required_if: list = kwargs.pop("not_required_if")

        assert self.not_required_if, "'not_required_if' parameter required"
        kwargs["help"] = (
            kwargs.get("help", "") + " Option is mutually exclusive with " + ", ".join(self.not_required_if) + "."
        ).strip()
        super().__init__(*args, **kwargs)

    def handle_parse_result(
        self, ctx: click.Context, opts: t.Mapping[str, t.Any], args: list[str]
    ) -> tuple[t.Any, list[str]]:
        """
        Overrides the `handle_parse_result` method of the `click.Option` class.

        This method checks if the current option is present in the provided options (`opts`)
        and if any of the mutually exclusive options are also present.
        If both the current option and a mutually exclusive option are present, it raises a `click.UsageError`.
        Otherwise, it returns the result of the parent `handle_parse_result` method.

        Args:
            ctx (click.Context): The Click context object.
            opts (dict): The dictionary of parsed options.
            args (list): The list of remaining arguments.

        Returns:
            The result of the parent `handle_parse_result` method.

        Raises:
            click.UsageError: If the current option and a mutually exclusive option are both present.
        """

        current_opt: bool = self.name in opts
        for mutex_opt in self.not_required_if:
            if mutex_opt in opts:
                if current_opt:
                    raise click.UsageError(
                        "Illegal usage: '" + str(self.name) + "' is mutually exclusive with " + str(mutex_opt) + "."
                    )
                self.prompt = None
        return super().handle_parse_result(ctx, opts, args)


def get_explorer_url(identifier: str | int, network: str, entity_type: ExplorerEntityType) -> str:
    """
    Returns a URL for exploring a specified type (transaction, asset, address) on the specified network.

    Args:
        identifier (str | int): The ID of the transaction, asset, or address to explore.
        network (str): The name of the network (e.g., "localnet", "testnet", "mainnet").
        entity_type (ExplorerEntityType): The type to explore (e.g., ExplorerEntityType.TRANSACTION,
        ExplorerEntityType.ASSET, ExplorerEntityType.ADDRESS).

    Returns:
        str: The URL for exploring the specified type on the specified network.

    Raises:
        ValueError: If the network or explorer type is invalid.
    """

    base_urls: dict[str, dict[str, str]] = {
        "testnet": {
            ExplorerEntityType.TRANSACTION.value: "https://testnet.algoexplorer.io/tx/",
            ExplorerEntityType.ASSET.value: "https://testnet.explorer.perawallet.app/assets/",
            ExplorerEntityType.ADDRESS.value: "https://testnet.algoexplorer.io/address/",
        },
        "mainnet": {
            ExplorerEntityType.TRANSACTION.value: "https://algoexplorer.io/tx/",
            ExplorerEntityType.ASSET.value: "https://explorer.perawallet.app/assets/",
            ExplorerEntityType.ADDRESS.value: "https://algoexplorer.io/address/",
        },
    }

    if network == "localnet":
        return f"https://app.dappflow.org/setnetwork?name=sandbox&redirect=explorer/{entity_type.value}/{identifier}/"

    if network not in base_urls:
        raise ValueError(f"Invalid network: {network}")

    if entity_type.value not in base_urls[network]:
        raise ValueError(f"Invalid explorer type: {entity_type}")

    return base_urls[network][entity_type.value] + str(identifier)
