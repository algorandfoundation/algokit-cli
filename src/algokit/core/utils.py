import re
import socket


def extract_version_triple(version_str: str) -> str:
    match = re.search(r"\d+\.\d+\.\d+", version_str)
    if not match:
        raise ValueError("Unable to parse version number")
    return match.group()


def is_minimum_version(system_version: str, minimum_version: str) -> bool:
    system_version_as_tuple = tuple(map(int, system_version.split(".")))
    minimum_version_as_tuple = tuple(map(int, minimum_version.split(".")))
    return system_version_as_tuple >= minimum_version_as_tuple


def is_network_available(host: str = "8.8.8.8", port: int = 53, timeout: float = 3.0) -> bool:
    """
    Check if internet is available by trying to establish a socket connection.
    """

    try:
        socket.setdefaulttimeout(timeout)
        socket.socket(socket.AF_INET, socket.SOCK_STREAM).connect((host, port))
        return True
    except OSError:
        return False
