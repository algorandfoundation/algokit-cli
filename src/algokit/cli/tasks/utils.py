# AlgoKit Tasks related utility functions

import logging
import os
import stat
import sys

import algosdk
import algosdk.encoding
import click
from algokit_utils import (
    Account,
    get_algod_client,
    get_algonode_config,
    get_default_localnet_config,
)
from algosdk.util import algos_to_microalgos, microalgos_to_algos

from algokit.cli.common.constants import AlgorandNetwork
from algokit.core.tasks.wallet import get_alias

logger = logging.getLogger(__name__)


def _validate_asset_balance(account_info: dict, asset_id: int, decimals: int, amount: int = 0) -> None:
    asset_record = next((asset for asset in account_info.get("assets", []) if asset["asset-id"] == asset_id), None)

    if not asset_record:
        raise click.ClickException("Account is not opted into the asset")

    if amount > 0 and asset_record["amount"] < amount:
        required = amount / 10**decimals
        available = asset_record["amount"] / 10**decimals
        raise click.ClickException(
            f"Insufficient asset balance in account, required: {required}, available: {available}"
        )


def _validate_algo_balance(account_info: dict, amount: int) -> None:
    if account_info.get("amount", 0) < amount:
        required = microalgos_to_algos(amount)  # type: ignore[no-untyped-call]
        available = microalgos_to_algos(account_info.get("amount", 0))  # type: ignore[no-untyped-call]
        raise click.ClickException(
            f"Insufficient Algos balance in account, required: {required} Algos, available: {available} Algos"
        )


def get_private_key_from_mnemonic() -> str:
    """
    Converts a mnemonic phrase into a private key.

    Returns:
        str: The private key generated from the mnemonic phrase.

    Raises:
        click.ClickException: If the entered mnemonic phrase is invalid.

    Example Usage:
        private_key = get_private_key_from_mnemonic()
        print(private_key)

    Inputs:
        None

    Flow:
        1. Prompts the user to enter a mnemonic phrase.
        2. Converts the entered mnemonic phrase into a private key using `algosdk.mnemonic.to_private_key`.
        3. Returns the private key.

    """

    mnemonic_phrase = click.prompt("Enter the mnemonic phrase (25 words separated by whitespace)", hide_input=True)
    try:
        return str(algosdk.mnemonic.to_private_key(mnemonic_phrase))  # type: ignore[no-untyped-call]
    except Exception as err:
        raise click.ClickException("Invalid mnemonic. Please provide a valid Algorand mnemonic.") from err


def load_algod_client(network: AlgorandNetwork) -> algosdk.v2client.algod.AlgodClient:
    """
    Returns an instance of the `algosdk.v2client.algod.AlgodClient` class for the specified network.

    Args:
        network (str): The network for which the `AlgodClient` instance needs to be loaded.

    Returns:
        algosdk.v2client.algod.AlgodClient: An instance of the `AlgodClient` class for the specified network.

    Raises:
        click.ClickException: If the specified network is invalid.
    """

    config_mapping = {
        AlgorandNetwork.LOCALNET: get_default_localnet_config("algod"),
        AlgorandNetwork.TESTNET: get_algonode_config("testnet", "algod", ""),
        AlgorandNetwork.MAINNET: get_algonode_config("mainnet", "algod", ""),
    }
    try:
        return get_algod_client(config_mapping[network])
    except KeyError as err:
        raise click.ClickException("Invalid network") from err


def get_asset_decimals(asset_id: int, algod_client: algosdk.v2client.algod.AlgodClient) -> int:
    """
    Retrieves the number of decimal places for a given asset.

    Args:
        asset_id (int): The ID of the asset for which the number of decimal places is to be retrieved. (0 for Algo)
        algod_client (algosdk.v2client.algod.AlgodClient): An instance of the AlgodClient class.

    Returns:
        int: The number of decimal places for the given asset.

    Raises:
        click.ClickException: If the asset info response is invalid.

    Example:
        asset_id = 123
        algod_client = algosdk.v2client.algod.AlgodClient("https://api.algoexplorer.io", "API_KEY")
        decimals = get_asset_decimals(asset_id, algod_client)
        print(decimals)
    """

    if asset_id == 0:
        return 6

    asset_info = algod_client.asset_info(asset_id)

    if not isinstance(asset_info, dict) or "params" not in asset_info or "decimals" not in asset_info["params"]:
        raise click.ClickException("Invalid asset info response")

    return int(asset_info["params"]["decimals"])


def validate_balance(
    algod_client: algosdk.v2client.algod.AlgodClient, account: Account | str, asset_id: int, amount: int = 0
) -> None:
    """
    Validates the balance of an account before an operation.

    Args:
        algod_client (algosdk.v2client.algod.AlgodClient): The AlgodClient object for
        interacting with the Algorand blockchain.
        account (Account | str): The account object.
        asset_id (int): The ID of the asset to be checked (0 for Algos).
        amount (int): The amount of Algos or asset for the operation. Defaults to 0 implying opt-in check only.

    Raises:
        click.ClickException: If any validation check fails.
    """
    address = account.address if isinstance(account, Account) else account
    account_info = algod_client.account_info(address)

    if not isinstance(account_info, dict):
        raise click.ClickException("Invalid account info response")

    if asset_id == 0:
        _validate_algo_balance(account_info, amount)
    else:
        decimals = get_asset_decimals(asset_id, algod_client)
        _validate_asset_balance(account_info, asset_id, decimals, amount)


def validate_address(address: str) -> None:
    """
    Check if a given address is a valid Algorand account address.

    Args:
        address (str): The address to be validated.

    Raises:
        click.ClickException: If the address is invalid.

    Returns:
        None
    """

    if not algosdk.encoding.is_valid_address(address):  # type: ignore[no-untyped-call]
        raise click.ClickException(f"`{address}` is an invalid account address")


def get_account_with_private_key(address: str) -> Account:
    """
    Retrieves an account object with the private key based on the provided address.

    Args:
        address (str): The address for which to retrieve the account object.

    Returns:
        Account: An account object with the address and private key.

    Raises:
        click.ClickException: If the address is not valid or if the alias does not exist or does not have a private key.
    """

    parsed_address = address.strip('"')

    try:
        validate_address(parsed_address)
        pk = get_private_key_from_mnemonic()
        return Account(address=parsed_address, private_key=pk)
    except click.ClickException as ex:
        alias_data = get_alias(parsed_address)

        if not alias_data:
            raise click.ClickException(f"Alias `{parsed_address}` alias does not exist.") from ex
        if not alias_data.private_key:
            raise click.ClickException(f"Alias `{parsed_address}` does not have a private key.") from ex

        return Account(address=alias_data.address, private_key=alias_data.private_key)


def get_address(address: str) -> str:
    """
    Validates the given address and returns it if valid. If the address is not valid,
    it checks if the address is an alias and returns the corresponding address if the alias exists.

    Args:
        address (str): The address to be validated or checked for an alias.

    Returns:
        str: The validated address or the corresponding address from an alias.

    Raises:
        click.ClickException: If the address is not valid and no corresponding alias exists.

    Example:
        address = get_address("ABCD1234")
        print(address)
    """

    parsed_address = address.strip('"')

    try:
        validate_address(parsed_address)
        return parsed_address
    except click.ClickException as ex:
        if len(parsed_address) == algosdk.constants.address_len:
            raise click.ClickException(f"`{parsed_address}` is an invalid account address") from ex

        alias_data = get_alias(parsed_address)

        if not alias_data:
            raise click.ClickException(f"Alias `{parsed_address}` alias does not exist.") from ex

        return alias_data.address


def stdin_has_content() -> bool:
    """
    Checks if there is content in the standard input.

    Returns:
        bool: True if there is content in the standard input, False otherwise.
    """

    mode = os.fstat(sys.stdin.fileno()).st_mode
    return stat.S_ISFIFO(mode) or stat.S_ISREG(mode)


def validate_account_balance_to_opt_in(
    algod_client: algosdk.v2client.algod.AlgodClient, account: Account, num_assets: int
) -> None:
    """
    Validates the balance of an account before opt in operation.
    Each asset requires 0.1 Algos to opt in.

    Args:
        algod_client (algosdk.v2client.algod.AlgodClient): The AlgodClient object for
        interacting with the Algorand blockchain.
        account (Account | str): The account object.
        num_assets (int): The number of the assets for opt in (0 for Algos).

    Raises:
        click.ClickException: If there is an insufficient fund in the account or account is not valid.
    """

    address = account.address if isinstance(account, Account) else account
    account_info = algod_client.account_info(address)

    if not isinstance(account_info, dict):
        raise click.ClickException("Invalid account info response")

    required_microalgos = num_assets * algos_to_microalgos(0.1)  # type: ignore[no-untyped-call]
    available_microalgos = account_info.get("amount", 0)
    if available_microalgos < required_microalgos:
        required_algo = microalgos_to_algos(required_microalgos)  # type: ignore[no-untyped-call]
        available_algos = microalgos_to_algos(available_microalgos)  # type: ignore[no-untyped-call]
        raise click.ClickException(
            f"Insufficient Algos balance in account to opt in, required: {required_algo} Algos, available:"
            f" {available_algos} Algos"
        )


def get_account_info(algod_client: algosdk.v2client.algod.AlgodClient, account_address: str) -> dict:
    account_info = algod_client.account_info(account_address)
    assert isinstance(account_info, dict)
    return account_info
