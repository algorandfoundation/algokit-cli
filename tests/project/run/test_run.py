import sys
from collections.abc import Callable
from pathlib import Path

import pytest
from _pytest.tmpdir import TempPathFactory
from pytest_mock import MockerFixture

from tests.utils.approvals import verify
from tests.utils.click_invoker import invoke
from tests.utils.proc_mock import ProcMock
from tests.utils.which_mock import WhichMock

DirWithAppSpecFactory = Callable[[Path], Path]

PYTHON_EXECUTABLE = sys.executable
# Escaping the python executable path for use in config files on Windows platforms
PYTHON_EXECUTABLE_ESCAPED = PYTHON_EXECUTABLE.replace("\\", "\\\\")


def _format_output(output: str) -> str:
    """
    Strips lines from the output that start with the specified string.

    Args:
        output (str): The output string to process.
        start_str (str): The string that identifies the start of lines to be stripped.

    Returns:
        str: The processed output with specified lines stripped.
    """
    output = "\n".join([line for line in output.split("\n") if not line.startswith("DEBUG")])
    return output.replace(PYTHON_EXECUTABLE_ESCAPED, "<sys.executable>").replace("\\", r"\\")


@pytest.fixture(autouse=True)
def _disable_animation(mocker: MockerFixture) -> None:
    mocker.patch("algokit.core.utils.animate", return_value=None)


@pytest.fixture()
def which_mock(mocker: MockerFixture) -> WhichMock:
    which_mock = WhichMock()
    mocker.patch("algokit.core.utils.shutil.which").side_effect = which_mock.which
    return which_mock


def _create_project_config(
    project_dir: Path, project_type: str, project_name: str, command: str, description: str
) -> None:
    """
    Creates a .algokit.toml configuration file in the specified project directory.

    Args:
        project_dir (Path): The directory of the project.
        project_type (str): The type of the project.
        project_name (str): The name of the project.
        command (str): The command associated with the project.
        description (str): A description of the project.
    """
    project_config = f"""
[project]
type = '{project_type}'
name = '{project_name}'

[project.run]
hello = {{ commands = ['{command}'], description = '{description}' }}
    """.strip()
    (project_dir / ".algokit.toml").write_text(project_config, encoding="utf-8")


def _create_workspace_project(
    *,
    workspace_dir: Path,
    projects: list[dict[str, str]],
    mock_command: bool = False,
    which_mock: WhichMock | None = None,
    proc_mock: ProcMock | None = None,
    custom_project_order: list[str] | None = None,
) -> None:
    """
    Creates a workspace project and its subprojects within the specified directory.

    Args:
        workspace_dir (Path): The directory of the workspace.
        projects (list[dict[str, str]]): A list of dictionaries, each representing a project with
        keys for directory, type, name, command, and description.
        mock_command (bool, optional): Indicates whether to mock the command. Defaults to False.
        which_mock (WhichMock | None, optional): The mock object for the 'which' command. Defaults to None.
        proc_mock (ProcMock | None, optional): The mock object for the process execution. Defaults to None.
        custom_project_order (list[str] | None, optional): Specifies a custom order for project execution.
        Defaults to None.
    """
    workspace_dir.mkdir()
    custom_project_order = custom_project_order if custom_project_order else ["contract_project", "frontend_project"]
    (workspace_dir / ".algokit.toml").write_text(
        f"""
[project]
type = 'workspace'
projects_root_path = 'projects'

[project.run]
hello = {custom_project_order}
        """.strip(),
        encoding="utf-8",
    )
    (workspace_dir / "projects").mkdir()
    for project in projects:
        project_dir = workspace_dir / "projects" / project["dir"]
        project_dir.mkdir()
        if mock_command and proc_mock and which_mock:
            resolved_mocked_cmd = which_mock.add(project["command"])
            proc_mock.set_output([resolved_mocked_cmd], ["picked " + project["command"]])

        _create_project_config(
            project_dir, project["type"], project["name"], project["command"], project["description"]
        )


@pytest.fixture()
def cwd_with_workspace_sequential(
    tmp_path_factory: TempPathFactory, which_mock: WhichMock, proc_mock: ProcMock
) -> Path:
    cwd = tmp_path_factory.mktemp("cwd") / "algokit_project"
    projects = [
        {
            "dir": "project1",
            "type": "contract",
            "name": "contract_project",
            "command": "command_a",
            "description": "Prints hello",
        },
        {
            "dir": "project2",
            "type": "frontend",
            "name": "frontend_project",
            "command": "command_b",
            "description": "Prints hello",
        },
    ]
    _create_workspace_project(
        workspace_dir=cwd, projects=projects, mock_command=True, which_mock=which_mock, proc_mock=proc_mock
    )

    return cwd


@pytest.fixture()
def cwd_with_workspace(tmp_path_factory: TempPathFactory, which_mock: WhichMock, proc_mock: ProcMock) -> Path:
    """
    Creates a standalone project with a single command.
    Single project is specified due to the fact that these are run concurrently,
    hence output stability is not guaranteed
    """

    cwd = tmp_path_factory.mktemp("cwd") / "algokit_project"
    projects = [
        {
            "dir": "project1",
            "type": "contract",
            "name": "contract_project",
            "command": "command_a",
            "description": "Prints hello",
        },
    ]
    _create_workspace_project(
        workspace_dir=cwd, projects=projects, mock_command=True, which_mock=which_mock, proc_mock=proc_mock
    )

    return cwd


@pytest.fixture()
def cwd_with_standalone(tmp_path_factory: TempPathFactory, which_mock: WhichMock, proc_mock: ProcMock) -> Path:
    cwd = tmp_path_factory.mktemp("cwd") / "algokit_project"
    cwd.mkdir()

    which_mock.add("command_a")
    proc_mock.set_output(["command_a"], ["picked command_a"])
    _create_project_config(cwd, "contract", "contract_project", "command_a", "Prints hello contracts")

    return cwd


def test_run_command_from_workspace_success(
    cwd_with_workspace: Path,
) -> None:
    """
    Verifies successful command execution within a workspace project.

    Args:
        cwd_with_workspace (Path): The path to the workspace directory.
    """
    result = invoke("project run hello", cwd=cwd_with_workspace)

    assert result.exit_code == 0
    verify(_format_output(result.output))


def test_run_command_from_workspace_sequential_success(cwd_with_workspace_sequential: Path) -> None:
    """
    Verifies successful sequential command execution within a workspace project.

    Args:
        cwd_with_workspace_sequential (Path): The path to the workspace directory.
    """
    result = invoke("project run hello", cwd=cwd_with_workspace_sequential)

    assert result.exit_code == 0
    verify(_format_output(result.output))


def test_run_command_from_standalone(cwd_with_standalone: Path) -> None:
    """
    Verifies successful command execution within a standalone project.

    Args:
        cwd_with_standalone (Path): The path to the standalone project directory.
    """
    result = invoke("project run hello", cwd=cwd_with_standalone)

    assert result.exit_code == 0
    verify(_format_output(result.output))


def test_run_command_from_workspace_filtered(cwd_with_workspace_sequential: Path) -> None:
    """
    Verifies successful command execution within a workspace project with filtering by project name.

    Args:
        cwd_with_workspace_sequential (Path): The path to the workspace directory.
    """
    result = invoke("project run hello --project-name 'contract_project'", cwd=cwd_with_workspace_sequential)

    assert result.exit_code == 0
    verify(_format_output(result.output))


def test_list_all_commands_in_workspace(cwd_with_workspace_sequential: Path) -> None:
    """
    Lists all commands available within a workspace project.

    Args:
        cwd_with_workspace_sequential (Path): The path to the workspace directory.
    """
    result = invoke("project run hello --list", cwd=cwd_with_workspace_sequential)

    assert result.exit_code == 0
    verify(_format_output(result.output))


def test_run_command_from_workspace_filtered_no_project(cwd_with_workspace_sequential: Path) -> None:
    """
    Verifies command execution within a workspace project when the specified project does not exist.

    Args:
        cwd_with_workspace_sequential (Path): The path to the workspace directory.
    """
    result = invoke("project run hello --project-name contract_project2", cwd=cwd_with_workspace_sequential)

    assert result.exit_code == 0
    verify(_format_output(result.output))


def test_run_command_from_workspace_resolution_error(
    tmp_path_factory: pytest.TempPathFactory,
) -> None:
    """
    Verifies the behavior when a command resolution error occurs within a workspace project.

    Args:
        tmp_path_factory (pytest.TempPathFactory): Pytest fixture to create temporary directories.
    """

    cwd = tmp_path_factory.mktemp("cwd") / "algokit_project"
    projects = [
        {
            "dir": "project2",
            "type": "frontend",
            "name": "frontend_project",
            "command": "failthiscommand",
            "description": "Prints hello",
        },
    ]
    _create_workspace_project(
        workspace_dir=cwd,
        projects=projects,
    )

    result = invoke("project run hello", cwd=cwd)

    assert result.exit_code == 1
    verify(_format_output(result.output))


def test_run_command_from_workspace_execution_error(
    tmp_path_factory: pytest.TempPathFactory,
) -> None:
    """
    Verifies the behavior when a command execution error occurs within a workspace project.

    Args:
        tmp_path_factory (pytest.TempPathFactory): Pytest fixture to create temporary directories.
    """
    cwd = tmp_path_factory.mktemp("cwd") / "algokit_project"
    projects = [
        {
            "dir": "project2",
            "type": "frontend",
            "name": "frontend_project",
            "command": PYTHON_EXECUTABLE_ESCAPED + ' -c "raise Exception()"',
            "description": "Prints hello",
        },
    ]
    _create_workspace_project(
        workspace_dir=cwd,
        projects=projects,
    )

    result = invoke("project run hello", cwd=cwd)

    assert result.exit_code == 1
    verify(_format_output(result.output))


def test_run_command_from_standalone_resolution_error(
    tmp_path_factory: pytest.TempPathFactory,
) -> None:
    """
    Verifies the behavior when a command resolution error occurs within a standalone project.

    Args:
        tmp_path_factory (pytest.TempPathFactory): Pytest fixture to create temporary directories.
    """
    cwd = tmp_path_factory.mktemp("cwd") / "algokit_project"
    projects = [
        {
            "dir": "project2",
            "type": "frontend",
            "name": "frontend_project",
            "command": "failthiscommand",
            "description": "Prints hello",
        },
    ]
    _create_workspace_project(
        workspace_dir=cwd,
        projects=projects,
    )

    result = invoke("project run hello", cwd=cwd / "projects" / "project2")

    assert result.exit_code == 1
    verify(_format_output(result.output))


def test_run_command_from_standalone_execution_error(tmp_path_factory: pytest.TempPathFactory) -> None:
    """
    Verifies the behavior when a command execution error occurs within a standalone project.

    Args:
        tmp_path_factory (pytest.TempPathFactory): Pytest fixture to create temporary directories.
    """
    cwd = tmp_path_factory.mktemp("cwd") / "algokit_project"
    cwd.mkdir()
    _create_project_config(
        cwd,
        "contract",
        "contract_project",
        PYTHON_EXECUTABLE_ESCAPED + ' -c "raise Exception()"',
        "Prints hello contracts",
    )

    result = invoke("project run hello", cwd=cwd)

    assert result.exit_code == 1
    verify(_format_output(result.output))


def test_run_command_from_workspace_partially_sequential(
    tmp_path_factory: TempPathFactory, which_mock: WhichMock, proc_mock: ProcMock
) -> None:
    """
    Verifies successful execution of commands in a partially sequential order within a workspace project.

    Args:
        tmp_path_factory (TempPathFactory): Pytest fixture to create temporary directories.
        which_mock (WhichMock): Mock object for the 'which' command.
        proc_mock (ProcMock): Mock object for process execution.
    """
    cwd = tmp_path_factory.mktemp("cwd") / "algokit_project"
    projects = []
    for i in range(1, 6):
        projects.append(
            {
                "dir": f"project{i}",
                "type": "contract",
                "name": f"contract_project_{i}",
                "command": f"hello{i}",
                "description": "Prints hello",
            }
        )
    _create_workspace_project(
        workspace_dir=cwd,
        projects=projects,
        mock_command=True,
        which_mock=which_mock,
        proc_mock=proc_mock,
        custom_project_order=["contract_project_1", "contract_project_4"],
    )

    result = invoke("project run hello", cwd=cwd)
    assert result.exit_code == 0
    order_of_execution = [line for line in result.output.split("\n") if line.startswith("✅")]
    assert "contract_project_1" in order_of_execution[0]
    assert "contract_project_4" in order_of_execution[1]


def test_run_command_from_standalone_pass_env(
    tmp_path_factory: TempPathFactory,
) -> None:
    """
    Verifies successful command execution within a standalone project with environment variables passed.

    Args:
        tmp_path_factory (TempPathFactory): Pytest fixture to create temporary directories.
    """
    cwd = tmp_path_factory.mktemp("cwd") / "algokit_project"
    cwd.mkdir()
    (cwd / "print_env.py").write_text('import os; print(os.environ.get("HELLO"))')

    _create_project_config(
        cwd,
        "contract",
        "contract_project",
        PYTHON_EXECUTABLE_ESCAPED + " print_env.py",
        "Prints hello contracts",
    )
    result = invoke("project run hello", cwd=cwd, env={"HELLO": "Hello World from env variable!"})

    assert result.exit_code == 0
    verify(_format_output(result.output))


def test_run_command_help_works_without_path_resolution(
    tmp_path_factory: TempPathFactory,
    which_mock: WhichMock,
    proc_mock: ProcMock,
) -> None:
    """
    Verifies that the help command works without path resolution.
    """

    cwd = tmp_path_factory.mktemp("cwd") / "algokit_project"
    projects = []
    for i in range(1, 6):
        projects.append(
            {
                "dir": f"project{i}",
                "type": "contract",
                "name": f"contract_project_{i}",
                "command": f"hello{i}",
                "description": "Prints hello",
            }
        )
    _create_workspace_project(
        workspace_dir=cwd,
        projects=projects,
        mock_command=False,
        which_mock=which_mock,
        proc_mock=proc_mock,
        custom_project_order=["contract_project_1", "contract_project_4"],
    )

    result = invoke("project run --help", cwd=cwd)

    assert result.exit_code == 0
    verify(_format_output(result.output))

    assert invoke("project run hello", cwd=cwd).exit_code == 1
