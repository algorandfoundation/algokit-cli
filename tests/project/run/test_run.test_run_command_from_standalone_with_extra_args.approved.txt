Running `hello` command in {current_working_directory}...
Command Executed: 'echo Hello'
Output: STDOUT
STDERR
Extra Args: 'extra args'
âœ… contract_project: 'echo Hello' executed successfully.
